### ISSUE TITLE
[windows] MSVC build broken

### ISSUE THREAD

**Issue by flriancu** [pocl/pocl#374]
Cheers, I'm having trouble building pocl on Windows using MSVC (and MSVC only).
I have the following:
Visual Studio 2013 Version 12.0.31101.00 Update 4
CMake 3.6.1 (the `bin/` is added to the path)
Python 2.7.11 (v2.7.11:6d1b6a68f775, Dec  5 2015, 20:32:19) [MSC v.1500 32 bit (Intel)] on win32
git version 2.7.3.windows.1

These are the issues I ran into:
- with `windows/setup_and_build_win64.sh` (executed from Git Bash)
  - www.open-mpi.com is now https
- as [it was failing](http://pastebin.com/raw/Vn9sTSY4), I changed line 641 in `pocl-playground\pocl\cmake\LLVM.cmake` to `LINK_LIBRARIES "${LLVM_LDFLAGS} ${LLVM_LIBFILES} ${LLVM_SYSLIBS}"`
- as [it was still failing](http://pastebin.com/raw/c9hw17UV), I compiled separately LLVM/Clang release_36 and added `bin/` to path; then deleted `pocl-playground\llvm` and `pocl-playground\llvm-build`;
- as [it was still failing](http://pastebin.com/raw/8ugiSyKx), I deleted the `LLVM_LINK_TEST` from `LLVM.cmake` (lines 600-654) and finally CMake [generated a MSVC project]().
- as [it would not compile](http://pastebin.com/raw/K9P9kLSL), I gave up:
  - it seems that MSVC has no `include_next` directive
  - (after fixing the headers with `include <../include/`whatever`.h>`) it appears some (lack of) casts triggers errors when building some sub-projects, as they are to be compiled as C++.

I do not have that much experience with LLVM/CMake. I tried to be as comprehensive as possible, tested on Windows 7 SP1 and Windows 10 (build before Anniversary Update), but in the end I'm still stuck.
Any help is greatly appreciated.

**Comment by flriancu** [comment]
Adding to it, there are headers that MSVC can't find, like `unistd.h` in `pocl-playground\pocl\lib\CL\devices` etc.
Was this written with MinGW in mind?

**Comment by pjaaskel** [comment]
> Was MSVC forgotten in these last 2 years?

That is probably close to the truth. I don't think anyone has cared for pocl on Windows for some time. It would be great if someone maintained it actively though knowing that Clang/LLVM on Windows has improved a lot! @elhigu did the initial port, he should be able to answer your questions if he has the time.

**Comment by pjaaskel** [comment]
I assume you noticed his notes in https://github.com/pocl/pocl/blob/master/README.Windows ?

**Comment by elhigu** [comment]
@flriancu I haven't been doing anything pocl related lately. Couple of years ago support was left in phase where all code did compile, but most of the tests were still failing. There is no Windows based CI build machine either, so most probably current master won't even compile on Windows.

**Comment by franz** [comment]
I tried to fix up Windows for 0.11 or 0.12 release (cant remember exactly), but by "fix up" i mean it did compile but didn't run anything (IIRC it was segfaulting in fopen() or some such thing). I tried to debug it, but apparently you need to link a program to a different 
msvc runtime dll to get debugging - and you also every other piece linked to that debugging dll (including clang) and i just couldn't get that mess working. So needs someone with real Windows expertise. Since then i guess nobody even thought about pocl + Windows, so probably we have tons of Windows-incompatible stuff by now.

**Comment by pjaaskel** [comment]
Please send a PR if you wish to readd support for Windows/MSVC.

**Comment by isuruf** [comment]
I built pthreads-win32 2.9.1, LLVM 6.0, Clang 6.0 with MSVC 2015 and hwloc 2.0.2 with clang-cl. I'm now trying to build pocl master with clang.

After fixing some cmake issues, I get an error about not finding some headers like stdlib.h. This is because the target is changed to `x86_64-pc` instead of `x86_64-windows-msvc` and clang doesn't find msvc headers nor understand msvc extensions like `__declspec`. I tried changing the triplet to `x86_64-windows-msvc` and it fails with "mangled name definition with same mangled name as another definition". I also noticed the comment about MicrosoftCXXMangler at https://github.com/pocl/pocl/blob/8c36c380afdfeebf9216627fa97a829f7a3d6cd5/cmake/LLVM.cmake#L149. Is this something that can be easily fixed on the clang side?
